sky/skylet/constants.py

```diff
@@ -73,12 +76,38 @@
 # AWS's Deep Learning AMI's default conda environment.
 CONDA_INSTALLATION_COMMANDS = (
     'which conda > /dev/null 2>&1 || '
-    '(wget -nc https://repo.anaconda.com/miniconda/Miniconda3-py310_23.5.2-0-Linux-x86_64.sh -O Miniconda3-Linux-x86_64.sh && '  # pylint: disable=line-too-long
+    '(wget -nc https://repo.anaconda.com/miniconda/Miniconda3-py310_23.11.0-2-Linux-x86_64.sh -O Miniconda3-Linux-x86_64.sh && '  # pylint: disable=line-too-long
     'bash Miniconda3-Linux-x86_64.sh -b && '
     'eval "$(~/miniconda3/bin/conda shell.bash hook)" && conda init && '
     'conda config --set auto_activate_base true); '
     'grep "# >>> conda initialize >>>" ~/.bashrc || conda init;')
 
+_sky_version = str(version.parse(sky.__version__))
+RAY_STATUS = f'RAY_ADDRESS=127.0.0.1:{SKY_REMOTE_RAY_PORT} ray status'
+RAY_SKYPILOT_INSTALLATION_COMMANDS = (
+    '(type -a python | grep -q python3) || '
+    'echo \'alias python=python3\' >> ~/.bashrc;'
+    '(type -a pip | grep -q pip3) || echo \'alias pip=pip3\' >> ~/.bashrc;'
+    'mkdir -p ~/sky_workdir && mkdir -p ~/.sky/sky_app;'
+    'source ~/.bashrc;'
+    # Backward compatibility for ray upgrade (#3248): do not upgrade ray if the
+    # ray cluster is already running, to avoid the ray cluster being restarted.
+    # NOTE: this will only work for the cluster with ray cluster on our latest
+    # ray port 6380, but those existing cluster launched before #1790 that has
+    # ray cluster on the default port 6379 will be upgraded and restarted.
+    f'{RAY_STATUS} || {{ pip3 list | grep "ray " | '
```

Comment:
Because we keep `ray==2.4.0` for a long time, and when we updated `ray` to `2.4.0` we assume all the users should be able to restart their cluster to adopt the new ray version.

Now, since some users may have existing cluster running the ray cluster with `2.4.0`, we have to avoid upgrading ray to make it backward compatible for those clusters, and ensure `sky exec` and `sky launch` on that cluster work.

Reply:
We need some comments here like, "We do this guard to avoid any Ray client-server version mismatch. Specifically: If existing ray cluster is an older version say 2.4, and we pip install new version say 2.9 wheels here, then subsequent `sky exec` (ray job submit) will have v2.9 vs. 2.4 mismatch, similarly this problem exists for `sky status -r` (ray status)."

Reply:
Added the comment! Thanks!